### DISCUSSION TITLE
GET request fires both on server side and client side.

### DISCUSSION THREAD

**Discussion by rmnkk** [nuxt/nuxt#29892] [Questions] [answered]
I want to fetch data from the API and set it up for the Pinia store. I use a [custom fetcher](https://nuxt.com/docs/guide/recipes/custom-usefetch#custom-usefetch) from the documentation example. But for some reason, the request fires twice: once on the server side and once on the client side.

Here is the [Stackblitz ](https://stackblitz.com/edit/nuxt-starter-og4ehc?file=pages%2Findex.vue)repo where you can check it. To test add .env file with this BASE_URL = https://jsonplaceholder.typicode.com
Also, I will duplicate files here for those who can't access Stackblitz:
I want to fetch data from the API and set it up for the Pinia store. I use a [custom fetcher](https://nuxt.com/docs/guide/recipes/custom-usefetch#custom-usefetch) from the documentation example. But, for some reason the request fires twice: first on the server side, second on the client side.
I created an api.ts plugin:

```
export default defineNuxtPlugin((nuxtApp) => {
  const config = useRuntimeConfig()
  const { session } = useUserSession()

  const api = $fetch.create({
    baseURL: config.public.API_BASE_URL,
    onRequest({ request, options, error }) {
      if (session.value?.token) {
        options.headers.set('Authorization', `Bearer ${session.value?.token}`)
      }
    },
    async onResponseError({ response }) {
      if (response.status === 401) {
        await nuxtApp.runWithContext(() => navigateTo('/login'))
      }
    },
  })

  return {
    provide: {
      api,
    },
  }
})
```
And useAPI composable:

```
import type { UseFetchOptions } from 'nuxt/app'

export function useAPI<T>(
  url: string | (() => string),
  options?: UseFetchOptions<T>,
) {
  return useFetch(url, {
    ...options,
    $fetch: useNuxtApp().$api as typeof $fetch,
  })
}
```
Pinia products store:

```
export const useProductsStore = defineStore('products', () => {
  const products = ref([])  
  const callCounter = ref(0)

  return {
    products,    
    callCounter,
  }
})
```
And to handle API requests I created useProducts composable. I use $api plugin instead of useAPI, because useFetch meant to be called directly in a setup function, plugin, or route middleware

```
export function useProducts() {
  const { $api } = useNuxtApp();
  const productsStore = useProductsStore();

  async function setProducts() {
    const data = await $api('/posts');
    console.log(data, 'data');
    productsStore.callCounter++;
    productsStore.products = data;
  }

  return {
    setProducts,
  };
}
```
On my page I use it like that. I use useAsyncData to prevent fetching on both the server and the client, but it doesn't work.

```
<script setup>
const { setProducts } = useProducts();
const productsStore = useProductsStore();
// setProducts()
const { data } = await useAsyncData(
  'products',
  async () => await setProducts()
);
</script>

<template>
  <main class="home">
    <div class="container">
      <article>
        {{ productsStore.callCounter }}
      </article>
    </div>
  </main>
</template>

<style scoped></style>
```
I tried to change $api for $fetch in useProducts composable, but the result is still the same: data fetched twice. What do I do wrong?

**Comment by logotip4ik** [comment, marked as answer]
useAsyncData should return something. This is needed for Nuxt, because it will check if the request was executed previously (check for return value, in this case it will be `undefined`). I think returning `true` from setProducts will be enough to fix the issue.

**Reply by rmnkk** [reply]
Thank you, addidng the `return` statement to the setProducts function has resolved the issue. I missed that part in the documentation:

> handler: an asynchronous function that must return a truthy value (for example, it should not be undefined or null) or the request may be duplicated on the client side
